pennylane_cirq/qsim_device.py

```diff
@@ -0,0 +1,111 @@
+"""
+import cirq
+import numpy as np
+
+try:
+    import qsimcirq
+except ImportError:
+    raise ImportError(
+        "QSim Cirq is needed for the QSim devices to work."
+        "\nIt can be installed using pip:"
+        "\n\npip install qsimcirq"
+    )
+
+from .simulator_device import SimulatorDevice
+from .cirq_device import CirqDevice
+
+
+class QSimDevice(SimulatorDevice):
+    r"""QSim device for PennyLane.
+    Args:
+        wires (int): the number of wires to initialize the device with
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. Shots need
+            to be >= 1. In analytic mode, shots indicates the number of entries
+            that are returned by ``device.sample``.
+        analytic (bool): indicates whether expectation values and variances should
+            be calculated analytically
+        qubits (List[cirq.Qubit]): A list of Cirq qubits that are used
+            as wires. The wire number corresponds to the index in the list.
+            By default, an array of ``cirq.LineQubit`` instances is created.
+    """
+    name = "QSim device for PennyLane"
+    short_name = "cirq.qsim"
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None):
+        super().__init__(wires, shots, analytic, qubits)
+        self.circuit = qsimcirq.QSimCircuit(cirq_circuit=cirq.Circuit())
+        self._simulator = qsimcirq.QSimSimulator()
+
+        self._initial_state = None
+        self._result = None
+        self._state = None
+
+
+class QSimhDevice(SimulatorDevice):
+    r"""QSimh device for PennyLane.
+    Args:
+        wires (int): the number of wires to initialize the device with
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. Shots need
+            to be >= 1. In analytic mode, shots indicates the number of entries
+            that are returned by ``device.sample``.
+        analytic (bool): indicates whether expectation values and variances should
+            be calculated analytically
+        qubits (List[cirq.Qubit]): A list of Cirq qubits that are used
+            as wires. The wire number corresponds to the index in the list.
+            By default, an array of ``cirq.LineQubit`` instances is created.
+        qsimh_options (dict): A dictionary with options for the QSimh simulator. See the `QSim
+            usage documentation <https://github.com/quantumlib/qsim/blob/master/docs/usage.md>`__
+            for further details.
+    """
+    name = "QSimh device for PennyLane"
+    short_name = "cirq.qsimh"
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None, qsimh_options=None):
+        super().__init__(wires, shots, analytic, qubits)
+        if qsimh_options is None:
+            qsimh_options = {
+                'k': [0],
+                'w': 0,
+                'p': 0,
+                'r': 1
```

Comment:
Are these options required or could we use a qsimh default?

Reply:
Also, would it make sense to make `qsimh_options` a required argument in this device? I'm not sure why those numbers above are good, so feel it might be good to force the user to decide.

Reply:
I'm quite sure that these options are required. I just put in the default recommended ones from the website here.

I guess having the options as a required argument makes a lot of sense (since they are required by qsimh). :thinking:

Reply:
Yes, I would vote to make `qsimh_options` a required arg, unless we're quite confident that those options make sense as a default.